File: elixir/allergies/lib/allergies.ex
defmodule Allergies do
  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags) do
    Stream.unfold(1, &{&1, &1 * 2})
    |> Enum.take(8)
    |> Enum.filter(&(Bitwise.band(flags, &1) > 0))
    |> Enum.map(&from_flag/1)
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item) do
    item in list(flags)
  end

  defp from_flag(0x01), do: "eggs"
  defp from_flag(0x02), do: "peanuts"
  defp from_flag(0x04), do: "shellfish"
  defp from_flag(0x08), do: "strawberries"
  defp from_flag(0x10), do: "tomatoes"
  defp from_flag(0x20), do: "chocolate"
  defp from_flag(0x40), do: "pollen"
  defp from_flag(0x80), do: "cats"
end

File: zig/allergies/allergies.zig
const std = @import("std");
const EnumSet = std.EnumSet;

pub const Allergen = enum(u32) {
    eggs = 1,
    peanuts = 2,
    shellfish = 4,
    strawberries = 8,
    tomatoes = 16,
    chocolate = 32,
    pollen = 64,
    cats = 128,
};

pub fn isAllergicTo(score: u8, allergen: Allergen) bool {
    return score & @intFromEnum(allergen) > 0;
}

const allergens = std.enums.values(Allergen);
pub fn initAllergenSet(score: usize) EnumSet(Allergen) {
    var result = EnumSet(Allergen).init(.{});
    inline for (allergens) |allergen| {
        const bit = @intFromEnum(allergen);
        if (score & bit > 0) {
            result.insert(allergen);
        }
    }
    return result;
}
